packages/autocertifier-client/src/AutoCertifierClient.ts

```diff
@@ -0,0 +1,172 @@
+    }
+
+    public stop(): void {
+        if (this.updateTimeout) {
+            clearTimeout(this.updateTimeout)
+            this.updateTimeout = undefined
+        }
+    }
+
+    private scheduleCertificateUpdate(expirationTimestamp: number): void {
+        if (this.updateTimeout) {
+            clearTimeout(this.updateTimeout)
+            this.updateTimeout = undefined
+        }
+        // update certificate 1 day before it expires
+        let updateIn = expirationTimestamp - Date.now()
+        if (updateIn > ONE_DAY) {
+            updateIn = updateIn - ONE_DAY
+        }
+
+        if (updateIn > MAX_INT_32) {
+            updateIn = MAX_INT_32
+        }
+
+        logger.info(updateIn + ' milliseconds until certificate update')
+        // TODO: use tooling from @streamr/utils to set the timeout with an abortController.
+        this.updateTimeout = setTimeout(this.checkSubdomainValidity, updateIn)
+    }
+
+    private createCertificate = async (): Promise<void> => {
+        const sessionId = await this.restClient.createSession()
+        let certifiedSubdomain: CertifiedSubdomain
+
+        this.ongoingSessions.add(sessionId)
+
+        try {
+            certifiedSubdomain = await this.restClient.createNewSubdomainAndCertificate(this.streamrWebSocketPort, sessionId)
+        } finally {
+            this.ongoingSessions.delete(sessionId)
+        }
+        const dir = path.dirname(this.subdomainPath)
+        // TODO: use async fs methods?
+        if (!fs.existsSync(dir)) {
+            fs.mkdirSync(dir, { recursive: true })
+        }
+        fs.writeFileSync(this.subdomainPath, JSON.stringify(certifiedSubdomain))
+        const certObj = forge.pki.certificateFromPem(certifiedSubdomain.certificate.cert)
+
+        const expirationTimestamp = certObj.validity.notAfter.getTime()
+        this.scheduleCertificateUpdate(expirationTimestamp)
+
+        this.emit('updatedSubdomain', certifiedSubdomain)
+    }
+
+    private updateCertificate = async (): Promise<void> => {
+        const sessionId = await this.restClient.createSession()
+        this.ongoingSessions.add(sessionId)
+
+        const oldSubdomain = JSON.parse(fs.readFileSync(this.subdomainPath, 'utf8')) as CertifiedSubdomain
+        const certifiedSubdomain = await this.restClient.updateCertificate(oldSubdomain.subdomain,
+            this.streamrWebSocketPort, oldSubdomain.token, sessionId)
+
+        this.ongoingSessions.delete(sessionId)
+
+        // TODO: use async fs methods?
+        fs.writeFileSync(this.subdomainPath, JSON.stringify(certifiedSubdomain))
+        const certObj = forge.pki.certificateFromPem(certifiedSubdomain.certificate.cert)
+
+        const expirationTimestamp = certObj.validity.notAfter.getTime()
+        this.scheduleCertificateUpdate(expirationTimestamp)
+
+        this.emit('updatedSubdomain', certifiedSubdomain)
+    }
+
+    // This method should be called by Streamr DHT whenever the IP address or port of the node changes
```

Comment:
This comment is not relevant in this package (the DHT is just one potential user of this package)